File: backend/components/metiers/threed/tests_centralises/utils/helpers/core/test_utils_helper_advanced.py
# test_utils_helper_advanced.py – Tests ultra avancés pour utils_helper.py (helpers/core)
from backend.components.metiers.threed.utils.helpers.core.utils_helper import format_date, is_object, deep_clone
import pytest
from backend.components.metiers.threed.utils.helpers.impl.advanced_impl_helper import validate_and_audit
from datetime import datetime

def test_validate_and_audit_core_valide():
    data = {'id': 10, 'user': 'CoreUser', 'email': 'core@example.com'}
    result = validate_and_audit(data)
    assert result['valid'] is True
    assert result['data']['user'] == 'anonyme'
    assert 'timestamp' in result['audit']

def test_validate_and_audit_core_invalide():
    result = validate_and_audit({})
    assert result['valid'] is False
    assert result['data'] == {}

def test_format_date_iso():
    d = datetime(2025, 6, 11, 12, 0, 0).isoformat()
    assert format_date(d) == '2025-06-11T12:00:00'

def test_is_object():
    assert is_object({'a': 1}) is True
    assert is_object([1, 2]) is False
    assert is_object(None) is False
    assert is_object('str') is False

def test_deep_clone():
    obj = {'a': 1, 'b': {'c': 2}}
    clone = deep_clone(obj)
    assert clone == obj
    assert clone is not obj
    clone['b']['c'] = 99
    assert obj['b']['c'] == 2

def test_deep_clone_empty():
    assert deep_clone({}) == {}

def test_format_date_type():
    d = datetime.now().isoformat()
    assert isinstance(format_date(d), str)

File: backend/components/metiers/threed/tests_centralises/utils/helpers/core/test_utils_helper_advanced.js
// test_utils_helper_advanced.js – Tests ultra avancés pour utils_helper.js (helpers/core)
const { validateAndAudit } = require('../../../../utils/helpers/impl/advanced_impl_helper');
const utils = require('../../../../utils/helpers/core/utils_helper');

describe('Helpers Core – validateAndAudit (Ultra avancé)', () => {
  it('valide et audite des données valides (core)', () => {
    const data = { id: 10, user: 'CoreUser', email: 'core@example.com' };
    const result = validateAndAudit(data);
    expect(result.valid).toBe(true);
    expect(result.data.user).toBe('anonyme');
    expect(result.audit).toHaveProperty('timestamp');
  });

  it('rejette les données invalides (core)', () => {
    const result = validateAndAudit({});
    expect(result.valid).toBe(false);
    expect(result.data).toEqual({});
  });
});

describe('Helpers Core – utils_helper (Ultra avancé)', () => {
  it('formatDate retourne une date ISO valide', () => {
    const d = new Date('2025-06-11T12:00:00Z');
    expect(utils.formatDate(d)).toBe('2025-06-11T12:00:00.000Z');
  });

  it('isObject détecte les objets purs', () => {
    expect(utils.isObject({ a: 1 })).toBe(true);
    expect(utils.isObject([1, 2])).toBe(false);
    expect(utils.isObject(null)).toBe(false);
    expect(utils.isObject('str')).toBe(false);
  });

  it('deepClone clone profondément un objet', () => {
    const obj = { a: 1, b: { c: 2 } };
    const clone = utils.deepClone(obj);
    expect(clone).toEqual(obj);
    expect(clone).not.toBe(obj);
    clone.b.c = 99;
    expect(obj.b.c).toBe(2);
  });

  it('robustesse: deepClone gère les objets vides', () => {
    expect(utils.deepClone({})).toEqual({});
  });

  it('auditabilité: formatDate retourne toujours une string', () => {
    expect(typeof utils.formatDate(new Date())).toBe('string');
  });
});
